File: test/problems/2-medium/3-Longest-Substring-Without-Repeating-Characters/solution.cpp
#include <string>
#include <unordered_set>

using namespace std;


int lengthOfLongestSubstring(string s) {
    unordered_set<char> substr_chars;
    int max_len = 0;
    int l = 0;
    int r = 0;
    while (r < s.size()) {
        while (substr_chars.find(s[r]) != substr_chars.end()) {
            substr_chars.erase(s[l]);
            l += 1;
        }
        substr_chars.insert(s[r]);
        r += 1;
        max_len = max(max_len, r-l);
    } 
    return max_len;
}

File: test/problems/2-medium/3-Longest-Substring-Without-Repeating-Characters/solution.java
import java.util.HashSet;
import java.util.Set;

public int lengthOfLongestSubstring(String s) {
    Set<Character> substrChars = new HashSet<>();
    int maxLen = 0;
    int l = 0;
    int r = 0;
    
    while (r < s.length()) {
        while (substrChars.contains(s.charAt(r))) {
            substrChars.remove(s.charAt(l));
            l++;
        }
        substrChars.add(s.charAt(r));
        r++;
        maxLen = Math.max(maxLen, r - l);
    }
    
    return maxLen;
}